### PR TITLE
JAMES-2215 Integration test about coherence of total and unread count…

### PR DESCRIPTION
**Pull request by aduprat**
…s on a shared folder

### CODE REVIEW

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+  Scenario: Sharer can read the total and unread counts on a shared folder
+    Given "alice@domain.tld" has a message "m1" in "shared" mailbox with subject "my test subject", content "testmail"
+    And "alice@domain.tld" has a message "m2" in "shared" mailbox with subject "my test subject 2", content "testmail 2"
+    When "alice@domain.tld" ask for mailboxes
+    Then a mailboxes answer is returned without error
+    And the list should contain 6 mailboxes
+    And the mailbox "shared" should be present
+    And the mailbox "shared" as 2 messages
+    And the mailbox "shared" as 2 unseen messages
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature]: s/as/has/ in both lines

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+  Scenario: Move message should update the total and the unread counts when asked by sharer
+    Given "alice@domain.tld" has a message "m1" in "INBOX" mailbox with subject "my test subject", content "testmail"
+    And "alice@domain.tld" moves "m1" to mailbox "shared" of user "alice@domain.tld"
+    When "alice@domain.tld" ask for mailboxes
+    Then a mailboxes answer is returned without error
+    And the list should contain 6 mailboxes
+    And the mailbox "shared" should be present
+    And the mailbox "shared" as 1 messages
+    And the mailbox "shared" as 1 unseen messages
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature]: if you move you should also check that counts in the original folder are 0, no?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+  Scenario: Move message should update the total and the unread counts when asked by sharer / sharee view
+    Given "alice@domain.tld" has a message "m1" in "INBOX" mailbox with subject "my test subject", content "testmail"
+    And "alice@domain.tld" moves "m1" to mailbox "shared" of user "alice@domain.tld"
+    When "bob@domain.tld" ask for mailboxes
+    Then a mailboxes answer is returned without error
+    And the list should contain 7 mailboxes
+    And the mailbox "shared" should be present
+    And the mailbox "shared" as 1 messages
+    And the mailbox "shared" as 1 unseen messages
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature]: idem

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+  Scenario: Move message should update the total and the unread counts when asked by sharee
+    Given "bob@domain.tld" has a message "m1" in "bobMailbox" mailbox with subject "my test subject", content "testmail"
+    And "bob@domain.tld" moves "m1" to mailbox "shared" of user "alice@domain.tld"
+    When "alice@domain.tld" ask for mailboxes
+    Then a mailboxes answer is returned without error
+    And the list should contain 6 mailboxes
+    And the mailbox "shared" should be present
+    And the mailbox "shared" as 1 messages
+    And the mailbox "shared" as 1 unseen messages
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature]: idem

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+  Scenario: Move message should update the total and the unread counts when asked by sharee / sharee view
+    Given "bob@domain.tld" has a message "m1" in "bobMailbox" mailbox with subject "my test subject", content "testmail"
+    And "bob@domain.tld" moves "m1" to mailbox "shared" of user "alice@domain.tld"
+    When "bob@domain.tld" ask for mailboxes
+    Then a mailboxes answer is returned without error
+    And the list should contain 7 mailboxes
+    And the mailbox "shared" should be present
+    And the mailbox "shared" as 1 messages
+    And the mailbox "shared" as 1 unseen messages
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature]: idem

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    Given "alice@domain.tld" has a mailbox "shared2"
+    And "alice@domain.tld" shares her mailbox "shared2" with "bob@domain.tld" with "lr" rights
+    And "bob@domain.tld" has a message "m1" in "bobMailbox" mailbox with subject "my test subject", content "testmail"
+    And "bob@domain.tld" moves "m1" to mailbox "shared2" of user "alice@domain.tld"
+    When "bob@domain.tld" ask for mailboxes
+    Then a mailboxes answer is returned without error
+    And the list should contain 8 mailboxes
+    And the mailbox "shared" should be present
+    And the mailbox "shared" as 0 messages
+    And the mailbox "shared" as 0 unseen messages
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature]: I think there is some missing test. When the message is shared, what happens when it is mark as read by the share or by sharer, from the point of you of the sharee, and of the sharer.

**Comment by mbaechler** [comment]
"coherence" is translated consistency in english

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMailboxesMethodStepdefs.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    @When("^\"([^\"]*)\" ask for mailboxes$")
+    public void getMailboxes(String user) throws Throwable {
+        userStepdefs.execWithUser(user, 
+                () -> httpClient.post("[[\"getMailboxes\", {}, \"#0\"]]"));
+    }
+
+    @Then("^a mailboxes answer is returned without error$")
+    public void mailboxesAnswer() throws Exception {
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMailboxesMethodStepdefs.java]: assertGetMailboxesOkStatus

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    Given a domain named "domain.tld"
+    And a user "alice@domain.tld"
+    And a user "bob@domain.tld"
+    And "alice@domain.tld" has a mailbox "INBOX"
+    And "bob@domain.tld" has a mailbox "bobMailbox"
+    And "alice@domain.tld" has a mailbox "shared"
+    And "alice@domain.tld" shares her mailbox "shared" with "bob@domain.tld" with "aeilrwt" rights
+
+  Scenario: Sharer can read the total and unread counts on a shared folder
+    Given "alice@domain.tld" has a message "m1" in "shared" mailbox with subject "my test subject", content "testmail"
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature]: In this case we don't care about the message content nor the messageIds, maybe we can write a "X has Y messages in Z mailbox" ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    And "alice@domain.tld" has a message "m2" in "shared" mailbox with subject "my test subject 2", content "testmail 2"
+    When "alice@domain.tld" ask for mailboxes
+    Then a mailboxes answer is returned without error
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature]: isn't it implicit ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    And the list should contain 6 mailboxes
+    And the mailbox "shared" should be present
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature]: I don't think that's the test intent

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    And "bob@domain.tld" has a mailbox "bobMailbox"
+    And "alice@domain.tld" has a mailbox "shared"
+    And "alice@domain.tld" shares her mailbox "shared" with "bob@domain.tld" with "aeilrwt" rights
+
+  Scenario: Sharer can read the total and unread counts on a shared folder
+    Given "alice@domain.tld" has a message "m1" in "shared" mailbox with subject "my test subject", content "testmail"
+    And "alice@domain.tld" has a message "m2" in "shared" mailbox with subject "my test subject 2", content "testmail 2"
+    When "alice@domain.tld" ask for mailboxes
+    Then a mailboxes answer is returned without error
+    And the list should contain 6 mailboxes
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature]: I don't think that's the test intent

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature (hunk 6 above)
Comment (mbaechler) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature]: I was about to ask the same thing (I was actually checking the ticket description but it's missing from it too)

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMailboxesMethodStepdefs.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    public void assertNumberOfMailboxes(int numberOfMailboxes) throws Exception {
+        assertThat(httpClient.jsonPath.<List<String>>read(ARGUMENTS + ".list")).hasSize(numberOfMailboxes);
+    }
+
+    @Then("^the mailbox \"([^\"]*)\" should be present$")
+    public void assertMailboxesNames(String mailboxName) throws Exception {
+        assertThat(httpClient.jsonPath.<List<String>>read(ARGUMENTS + ".list[*].name")).contains(mailboxName);
+    }
+
+    @Then("^the mailbox \"([^\"]*)\" as (\\d+) messages$")
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMailboxesMethodStepdefs.java]: s/ messages / (?:message|messages)

Right?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMailboxesMethodStepdefs.java [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    public void assertNumberOfMessages(String mailboxName, int numberOfMessages) throws Exception {
+        assertThat(httpClient.jsonPath.<List<Integer>>read(ARGUMENTS + ".list[?].totalMessages",
+                filter(where("name").is(mailboxName))))
+            .containsOnly(numberOfMessages);
+    }
+
+    @Then("^the mailbox \"([^\"]*)\" as (\\d+) unseen messages$")
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMailboxesMethodStepdefs.java]: s/ messages / (?:message|messages)

Right?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+    And the mailbox "shared" should be present
+    And the mailbox "shared" as 2 messages
+    And the mailbox "shared" as 2 unseen messages
+
+  Scenario: Sharee can read the total and unread counts on a shared folder
+    Given "alice@domain.tld" has a message "m1" in "shared" mailbox with subject "my test subject", content "testmail"
+    And "alice@domain.tld" has a message "m2" in "shared" mailbox with subject "my test subject 2", content "testmail 2"
+    When "bob@domain.tld" ask for mailboxes
+    Then a mailboxes answer is returned without error
+    And the list should contain 7 mailboxes
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature]: That 7 seems a bit magical. Maybe it is pointless in that test.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature (hunk 6 above)
Comment (chibenwa) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature]: We also miss:

 - Copying a Seen flag with lri rights should update total count, but not unseen count.
 - Does "l" right allow me to see "total" and "unseen"

**Comment by chibenwa** [comment]
```
[477c57bd44c996218fcd25b341afbff9061b09a7] [ERROR] Failures: 
[477c57bd44c996218fcd25b341afbff9061b09a7] [ERROR]   CassandraMailboxManagerTest>MailboxManagerTest.getMailboxCountersShouldReturnStoredValueWhenReadRight:919 expected:<...el.MailboxCounters@3[e]1> but was:<...el.MailboxCounters@3[c]1>
[477c57bd44c996218fcd25b341afbff9061b09a7] [INFO] 
[477c57bd44c996218fcd25b341afbff9061b09a7] [ERROR] Tests run: 715, Failures: 1, Errors: 0, Skipped: 5
```

Investigating...

**Comment by chibenwa** [comment]
Oh crap I found it, I need https://github.com/linagora/james-project/pull/1121/commits/ae3030f42bff17c738006fbacfec073446ff7bb2

**Comment by aduprat** [comment]
:+1: for me

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature [hunk 15]
```diff
@@ -226,7 +225,7 @@ Feature: GetMailboxes method
     Then the mailbox "shared2" has 1 message
     And the mailbox "shared2" has 1 unseen message
 
-  Scenario: Set flags by sharee should not update unseen count when no rights and read by sharee
+  Scenario: As sharee read a message it should not update unseen count when no rights and read by sharee
```
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMailboxes.feature]: We can't use `When` in scenario names...

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 16]
```diff
@@ -538,6 +522,27 @@ public MetaData getMetaData(boolean resetRecent, MailboxSession mailboxSession,
         return new MailboxMetaData(recent, permanentFlags, uidValidity, uidNext, highestModSeq, messageCount, unseenCount, firstUnseen, isWriteable(mailboxSession), isModSeqPermanent(mailboxSession), resolvedAcl);
     }
 
+    private MetaData sensibleInformationFreeMetaData(MailboxSession mailboxSession, MailboxACL resolvedAcl) throws MailboxException {
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: This seems to be even possible to make it a static constant in MailboxMetaData class (or at least a static method of this class)

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java [hunk 17]
```diff
@@ -360,14 +359,14 @@ public void appendMessageWithFlags(String username, String messageName, String m
...
+        appendMessage(messageName, mailbox, StringListToFlags.fromFlagList(flagList));
     }
 
-    private void appendMessage(String messageName, Flags flags) throws Exception {
+    private void appendMessage(String messageName, String mailbox, Flags flags) throws Exception {
         ZonedDateTime dateTime = ZonedDateTime.parse("2014-10-30T14:12:00Z");
         boolean isRecent = flags.contains(Flags.Flag.RECENT);
         MessageId id = mainStepdefs.mailboxProbe.appendMessage(userStepdefs.getConnectedUser(),
-            MailboxPath.forUser(userStepdefs.getConnectedUser(), DefaultMailboxes.INBOX),
+            MailboxPath.forUser(userStepdefs.getConnectedUser(), mailbox),
```
[the remark this answers is not in the document]
Comment (aduprat) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: I agree with you, but it's leading to some refactoring which is not really needed in this review.
For example, it's breaking `MailboxPath.getHierarchyLevels` method...

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/MailboxMetaData.java [hunk 18]
```diff
@@ -26,14 +26,39 @@
...
 import org.apache.james.mailbox.model.MailboxACL;
 
+import com.google.common.collect.ImmutableList;
+
 /**
  * Describes the current state of a mailbox.
  */
 public class MailboxMetaData implements MessageManager.MetaData {
 
+    public static MetaData sensibleInformationFree(MailboxACL resolvedAcl, long uidValidity, boolean writeable, boolean modSeqPermanent) throws MailboxException {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/MailboxMetaData.java]: it should return MailboxMetadata.

**Comment by aduprat** [comment]
test this please

**Comment by quynhn** [comment]
test this please

**Comment by chibenwa** [comment]
```
[90780a35b59fbcda5216c4f38f4d17cfdeb034d8] [INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 21.841 s - in org.apache.james.jmap.cassandra.CassandraUserProvisionningConcurrencyTest
[90780a35b59fbcda5216c4f38f4d17cfdeb034d8] [INFO] 
[90780a35b59fbcda5216c4f38f4d17cfdeb034d8] [INFO] Results:
[90780a35b59fbcda5216c4f38f4d17cfdeb034d8] [INFO] 
[90780a35b59fbcda5216c4f38f4d17cfdeb034d8] [ERROR] Errors: 
[90780a35b59fbcda5216c4f38f4d17cfdeb034d8] [ERROR]   org.apache.james.mailbox.exception.MailboxNotFoundException: #private:alice@domain.tld:inbox can not be found
[90780a35b59fbcda5216c4f38f4d17cfdeb034d8] [ERROR]   org.apache.james.mailbox.exception.MailboxNotFoundException: #private:alice@domain.tld:inbox can not be found
[90780a35b59fbcda5216c4f38f4d17cfdeb034d8] [ERROR]   org.apache.james.mailbox.exception.MailboxNotFoundException: #private:alice@domain.tld:inbox can not be found
[90780a35b59fbcda5216c4f38f4d17cfdeb034d8] [ERROR]   org.apache.james.mailbox.exception.MailboxNotFoundException: #private:alice@domain.tld:inbox can not be found
```

**Comment by chibenwa** [comment]
test this please

**Pull request merged**